runtime/sys.c

```diff
@@ -434,27 +434,54 @@ CAMLprim value caml_sys_unsafe_getenv(value var)
   return val;
 }
 
-CAMLprim value caml_sys_getenv(value var)
+/* Returns either a string or Val_none on error (i.e. the tag is used to encode
+   error conditions */
+static value sys_getenv(value var)
 {
   char_os * res, * p;
   value val;
 
-  if (! caml_string_is_c_safe(var)) caml_raise_not_found();
+  if (! caml_string_is_c_safe(var))
+    return Val_none;
   p = caml_stat_strdup_to_os(String_val(var));
 #ifdef _WIN32
   res = caml_win32_getenv(p);
 #else
   res = caml_secure_getenv(p);
 #endif
   caml_stat_free(p);
-  if (res == 0) caml_raise_not_found();
+  if (res == 0)
+    return Val_none;
   val = caml_copy_string_of_os(res);
 #ifdef _WIN32
   caml_stat_free(res);
 #endif
   return val;
 }
 
+CAMLprim value caml_sys_getenv(value var)
+{
+  value res = sys_getenv(var);
+
+  if (Is_none(res))
+    caml_raise_not_found();
+
+  return res;
+}
+
+CAMLprim value caml_sys_getenv_opt(value var)
+{
+  CAMLparam0(); /* var is used before allocation */
+  CAMLlocal1(res); /* root for the result, as it may go via caml_alloc_some */
```

Comment:
I believe that `CAMLlocal1` is not actually needed here, but it doesn't harm in any case.

Reply:
Oh, indeed it's not - I'd been doing something else at the same time/place (with https://github.com/dra27/ocaml/commit/a131afbe79dedbb1465922d90047be47fae0b281) where that did matter, and I've obviously confused myself!

I think it does look strange to have them being different with the wrong explanation. FWIW, I'd prefer to remove it, rather than add it to `caml_sys_getenv`, but I don't mind - what do you think?

Reply:
Removing it sounds fine!

Reply:
It is not obvious to me that not-rooting `res` across `caml_alloc_some` is safe. Doesn't each caller need to root the value?

Reply:
`caml_alloc_some` roots its argument (because it allocates)

Reply:
Yes, but this does not imply that the value `res` will be updated if the underlying block gets moved, just that the block will be kept alive by the GC. (But maybe I'm just confusing myself?)

Reply:
Ah, indeed, I am confusing myself: the argument which was previously `res` may get moved, but this is okay because we are never using it ever again and only care about the return value of `caml_alloc_some`. So I think this is safe. It would be wrong to try to use the *previous* `res` again, despite the fact that `caml_alloc_some` roots its argument.

Reply:
I think the name being the same is what's confusing (having also realised that the other commit I referred to also does not _require_ a root!). `res` holds the single allocated value from `sys_getenv` (pointer a string block). If that string is moved during the allocation in `caml_alloc_some`, then _our_ `res` value will not be updated, but the `v` argument in `caml_alloc_some` will be, meaning the new location of the string-block gets stored in the newly-allocated some-block. On return from `caml_alloc_some`, `res` immediately gets overwritten with the pointer to the new some-block, and there are then no further allocations.